### PR TITLE
Improvements for Edit-with-Emacs feature

### PR DESCRIPTION
**Pull request by agzam**
- Simplified and streamlined it to work not only with Spacemacs
- Made it so it doesn't block the thread on emacsclient call
- Added a hook for customization

### CODE REVIEW

**Comment by eccentric-j** [comment]
I'll try it out tomorrow before I approve or request changes but the code and docs do look solid. 

One part that could be expanded on though is how to install the package, or if we can make it easier to install the package. 

I was able to assemble a spacemacs layer so maybe we can include that somehow?

```elisp
;;; packages.el --- spacehammer layer packages file for Spacemacs.
;;
;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
;;
;; URL: https://github.com/syl20bnr/spacemacs
;;
;; This file is not part of GNU Emacs.
;;
;;; License: GPLv3

;;; Commentary:

;; See the Spacemacs documentation and FAQs for instructions on how to implement
;; a new layer:
;;
;;   SPC h SPC layers RET
;;
;;
;; Briefly, each package to be installed or configured by this layer should be
;; added to `spacehammer-packages'. Then, for each package PACKAGE:
;;
;; - If PACKAGE is not referenced by any other Spacemacs layer, define a
;;   function `spacehammer/init-PACKAGE' to load and initialize the package.

;; - Otherwise, PACKAGE is already referenced by another Spacemacs layer, so
;;   define the functions `spacehammer/pre-init-PACKAGE' and/or
;;   `spacehammer/post-init-PACKAGE' to customize the package as it is loaded.

;;; Code:

(defconst spacehammer-packages
  '((spacehammer :location
                 (recipe
                  :fetcher file
                  :path "~/.hammerspoon/")))
  "The list of Lisp packages required by the spacehammer layer.

Each entry is either:

1. A symbol, which is interpreted as a package to be installed, or

2. A list of the form (PACKAGE KEYS...), where PACKAGE is the
    name of the package to be installed or loaded, and KEYS are
    any number of keyword-value-pairs.

    The following keys are accepted:

    - :excluded (t or nil): Prevent the package from being loaded
      if value is non-nil

    - :location: Specify a custom installation location.
      The following values are legal:

      - The symbol `elpa' (default) means PACKAGE will be
        installed using the Emacs package manager.

      - The symbol `local' directs Spacemacs to load the file at
        `./local/PACKAGE/PACKAGE.el'

      - A list beginning with the symbol `recipe' is a melpa
        recipe.  See: https://github.com/milkypostman/melpa#recipe-format")

(defun spacehammer/init-spacehammer ()
  (use-package spacehammer
    :demand t
    :config
    (progn
      (spacemacs/transient-state-register-add-bindings 'zoom-frm
        '(("h" (spacehammer/move-frame-one-display "West"))
          ("l" (spacehammer/move-frame-one-display "East"))
          ("n" (spacehammer/move-frame-one-display "North"))
          ("p" (spacehammer/move-frame-one-display "South"))))

      (define-advice spacemacs/zoom-frm-transient-state/nil
          (:after () fix-frame-after-zoom-frm-transient-state)
        (spacehammer/fix-frame)))))

;;; packages.el ends here
```

**Comment by agzam** [comment]
Maybe at some point, I can try deploying the package to MELPA. Spacemacs layers are unfortunately not very modular in the sense that you cannot "import a layer" from a repository. Perhaps I should extend the documentation to include instructions for making a custom Spacemacs layer, or adding necessary pieces to an existing one.

**Comment by agzam** [comment]
@eccentric-j I've added a couple of paragraphs. Let me know what you think.

File: docs/emacs.org [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        ;; and the recipe would have to be something like this:
+
+        (defconst my-layer-packages
+          '((spacehammer :location local)))
+
+        ;; if you'd like to use the same Spacemacs config on different machines that
+        ;; aren't Macs, and you don't want it to complain about not finding the package
+        ;; (since Hammerspoon is not there):
+
+        (defconst my-layer-packages `(,(when (eq system-type 'darwin)
```
Comment (eccentric-j) [inline docs/emacs.org:70]: Is the `(,` intentional?

File: docs/emacs.org [hunk 2]
```diff
@@ -0,0 +1,106 @@
+* Edit with Emacs
+
+  Any text, in just any app can be edited using Emacs. The idea is simple - you press dedicated key-combo (set to =Cmd+Ctrl+O=), Spacehammer copies existing text, then calls ~emacsclient~, then invokes a function that opens a buffer and pastes the text into it. After you're done editing the text in Emacs, you press =C-c C-c= - it grabs the text, switches back to the program, pastes the text back in there. It works surprisingly well.
```
Comment (eccentric-j) [inline docs/emacs.org:3]: Ahhh now I see how this is supposed to work. Thanks!

File: docs/emacs.org (hunk 1 above)
Comment (agzam) [reply docs/emacs.org:70]: yes, it's a backquote marker https://www.gnu.org/software/emacs/manual/html_node/elisp/Backquote.html

**Comment by agzam** [comment]
Thanks @eccentric-j 

I have added a .gif demo, can you take a look? https://github.com/agzam/spacehammer/blob/1bd17bc42414372dbf3249f1c320def9ff77241b/docs/emacs.org

It's a bit heavy > 1Mb, you think it's worth keeping it in the repo?

Also I still need an approval to merge it.

**Pull request merged**